### PR TITLE
Add new ICL kwargs in eval.py and long_context yamls

### PR DESCRIPTION
**Pull request by maxisawesome**
Adds new kwargs included in the icl_refactor
Namely:
- generation_kwargs
- hf_parsing_map
- hf_loading_vars

Required CoT criteria are there as well.
- do_normalization
- early_stopping_criteria

I also added `fewshot_random_seed` to be settable from the root level of the yaml, i.e. the same as `max_seq_len`.

`mpt-7b-8k`  results from `test-loooooong-context-vdFLBW`:
```
Printing complete results for all models
| Category   | Benchmark              | Subtask   |   Accuracy | Number few shot   | Model     |
|:-----------|:-----------------------|:----------|-----------:|:------------------|:----------|
|            | kv_pairs_beginning_2k  |           |   0.946667 | 0-shot            | mpt-7b-8k |
|            | kv_pairs_middle_2k     |           |   0.936667 | 0-shot            | mpt-7b-8k |
|            | kv_pairs_end_2k        |           |   0.95     | 0-shot            | mpt-7b-8k |
|            | kv_pairs_beginning_4k  |           |   0        | 0-shot            | mpt-7b-8k |
|            | kv_pairs_middle_4k     |           |   0.383333 | 0-shot            | mpt-7b-8k |
|            | kv_pairs_end_4k        |           |   0.92     | 0-shot            | mpt-7b-8k |
|            | kv_pairs_beginning_8k  |           |   0        | 0-shot            | mpt-7b-8k |
|            | kv_pairs_middle_8k     |           |   0        | 0-shot            | mpt-7b-8k |
|            | kv_pairs_end_8k        |           |   0.673333 | 0-shot            | mpt-7b-8k |
|            | wikiqa_2k              |           |   0.149533 | 0-shot            | mpt-7b-8k |
|            | wikiqa_4k              |           |   0.149533 | 0-shot            | mpt-7b-8k |
|            | wikiqa_8k              |           |   0.149533 | 0-shot            | mpt-7b-8k |
|            | hotpotqa_beginning_2k  |           |   0.3625   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_middle_2k     |           |   0.3715   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_end_2k        |           |   0.379    | 0-shot            | mpt-7b-8k |
|            | hotpotqa_beginning_4k  |           |   0.158    | 0-shot            | mpt-7b-8k |
|            | hotpotqa_middle_4k     |           |   0.2595   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_end_4k        |           |   0.4075   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_beginning_8k  |           |   0.1585   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_middle_8k     |           |   0.163    | 0-shot            | mpt-7b-8k |
|            | hotpotqa_end_8k        |           |   0.342    | 0-shot            | mpt-7b-8k |
|            | hotpotqa_beginning_16k |           |   0.155    | 0-shot            | mpt-7b-8k |
|            | hotpotqa_beginning_32k |           |   0.1505   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_beginning_64k |           |   0.156    | 0-shot            | mpt-7b-8k |
|            | hotpotqa_middle_16k    |           |   0.1535   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_middle_32k    |           |   0.1595   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_middle_64k    |           |   0.16     | 0-shot            | mpt-7b-8k |
|            | hotpotqa_end_16k       |           |   0.247    | 0-shot            | mpt-7b-8k |
|            | hotpotqa_end_32k       |           |   0.2055   | 0-shot            | mpt-7b-8k |
|            | hotpotqa_end_64k       |           |   0.1885   | 0-shot            | mpt-7b-8k |
|            | kv_pairs_beginning_16k |           |   0.155    | 0-shot            | mpt-7b-8k |
|            | kv_pairs_beginning_32k |           |   0.1505   | 0-shot            | mpt-7b-8k |
|            | kv_pairs_beginning_64k |           |   0.156    | 0-shot            | mpt-7b-8k |
|            | kv_pairs_middle_16k    |           |   0.1535   | 0-shot            | mpt-7b-8k |
|            | kv_pairs_middle_32k    |           |   0.1595   | 0-shot            | mpt-7b-8k |
|            | kv_pairs_middle_64k    |           |   0.16     | 0-shot            | mpt-7b-8k |
|            | kv_pairs_end_16k       |           |   0.247    | 0-shot            | mpt-7b-8k |
|            | kv_pairs_end_32k       |           |   0.2055   | 0-shot            | mpt-7b-8k |
|            | kv_pairs_end_64k       |           |   0.1885   | 0-shot            | mpt-7b-8k |
```

### CODE REVIEW

**Review by dakinggg — CHANGES_REQUESTED** [review]
Must wait for composer release

File: scripts/eval/eval.py
Comment (dakinggg) [inline scripts/eval/eval.py:306]: does this need to be a new top level arg? I feel like it should just the seed top level arg that already exists, and then individual icl tasks hsould be able to override it if needed. wdyt

File: scripts/eval/yamls/hf_eval.yaml
Comment (dakinggg) [inline scripts/eval/yamls/hf_eval.yaml:15]: this should be mixed

File: scripts/eval/yamls/long_context_eval_8k.yaml
Comment (dakinggg) [inline scripts/eval/yamls/long_context_eval_8k.yaml:11]: mixed

File: scripts/eval/yamls/long_context_eval_8k.yaml
Comment (dakinggg) [inline scripts/eval/yamls/long_context_eval_8k.yaml:1]: is this yaml really needed? its just the hf eval one with a different tasks/gauntlet file

File: scripts/eval/yamls/long_context_tasks.yaml
Comment (dakinggg) [inline scripts/eval/yamls/long_context_tasks.yaml:4]: we should move these datasets into the mosaicml org, or find a different way to distribute them then via your personal HF account

File: setup.py
Comment (dakinggg) [inline setup.py:53]: this shouldn't need changing after merge

File: scripts/eval/yamls/long_context_eval_8k.yaml
Comment (maxisawesome) [reply scripts/eval/yamls/long_context_eval_8k.yaml:1]: It's just meant to be a designation for the long context tasks as they're a set of tasks that are explicitly not the gauntlet.

File: scripts/eval/eval.py
Comment (maxisawesome) [reply scripts/eval/eval.py:306]: I don't see random_seed anywhere else? Can you point me there? 
Per task would be better imo, some tasks might vary with different random seed settings more than others and that would be more controllable.

File: scripts/eval/yamls/long_context_eval_8k.yaml
Comment (maxisawesome) [reply scripts/eval/yamls/long_context_eval_8k.yaml:1]: But no, it's not strictly need and can remove.

Comment (dakinggg) [reply scripts/eval/yamls/long_context_eval_8k.yaml:1]: Yeah, I think just remove this. Want to keep the examples not too cluttered. Since its just a couple lines changed from the main hf eval yaml

File: scripts/eval/eval.py
Comment (dakinggg) [reply scripts/eval/eval.py:306]: https://github.com/mosaicml/llm-foundry/blob/2e0a845863dc18ad8ddac71fc7f18a9940989891/scripts/eval/eval.py#L289

I'd suggest that you default to this seed, but allow overriding in the individual task configs

Comment (maxisawesome) [reply scripts/eval/eval.py:306]: ffs I searched random and random_seed but not seed lol
what you have is much better than this big pass through

Comment (dakinggg) [reply scripts/eval/eval.py:306]: lol np, there should be docs and a dataclass :P

File: scripts/eval/yamls/long_context_eval_8k.yaml
Comment (maxisawesome) [reply scripts/eval/yamls/long_context_eval_8k.yaml:1]: Should I also remove the several long context gauntlet examples as well then?

Comment (dakinggg) [reply scripts/eval/yamls/long_context_eval_8k.yaml:1]: I think if y'all intend to have a "long context gauntlet" its good to include, otherwise can remove. I'm less opinionated on this one though and whatever you prefer is ok.

Comment (maxisawesome) [reply scripts/eval/yamls/long_context_eval_8k.yaml:1]: we've used them to identify pitfalls in the model, and explicitly organizing by both section and length is important. I'm gonna keep them, but fix them up a bit

**Review by dakinggg — APPROVED** [review]

**Comment by bmosaicml** [comment]
👨‍🍳  great work Max! Do you think the long-context gauntlet would justify it's own readme, akin to EVAL_GAUNTLET.md?

**Review by bmosaicml — APPROVED** [review]
Looks good, would you mind running a small model through the long context gauntlet and posting the results?

**Comment by maxisawesome** [comment]
> 👨‍🍳 great work Max! Do you think the long-context gauntlet would justify it's own readme, akin to EVAL_GAUNTLET.md?

I added it to the EVAL_GAUNTLET.md because I felt like it didn't deserve a full readme. Can redo if you'd like.

**Pull request merged**
